### PR TITLE
Implemented the method : getTimeBefore(Date beforeTime)

### PR DESCRIPTION
**Pull request by CShawn**
Implemented the method : getTimeBefore(Date beforeTime) in CronExpression.java

### CODE REVIEW

File: quartz-core/src/main/java/org/quartz/CronExpression.java [hunk 1]
```diff
@@ -1589,9 +1590,409 @@ protected void setCalendarHour(Calendar cal, int hour) {
      * NOT YET IMPLEMENTED: Returns the time before the given time
      * that the <code>CronExpression</code> matches.
      */ 
-    public Date getTimeBefore(Date endTime) { 
-        // FUTURE_TODO: implement QUARTZ-423
-        return null;
+    public Date getTimeBefore(Date beforeTime) {
```
Comment (maheshwarigagan) [inline quartz-core/src/main/java/org/quartz/CronExpression.java]: Can you add unit test cases for this method?

File: quartz-core/src/main/java/org/quartz/CronExpression.java [hunk 2]
```diff
@@ -1589,9 +1590,409 @@ protected void setCalendarHour(Calendar cal, int hour) {
...
+                            cl.set(Calendar.MONTH, mon - 1 - 1);
+                        }
+                        continue;
+                    } else if (daysToAdd < 0) { // are we swithing days?
+                        cl.set(Calendar.SECOND, 59);
+                        cl.set(Calendar.MINUTE, 59);
+                        cl.set(Calendar.HOUR_OF_DAY, 23);
+                        cl.set(Calendar.DAY_OF_MONTH, day + daysToAdd);
+                        cl.set(Calendar.MONTH, mon - 1);
+                        // '- 1' because calendar is 0-based for this field,
```
Comment (maheshwarigagan) [inline quartz-core/src/main/java/org/quartz/CronExpression.java]: This code can be moved to a private method, will remove the redundancy very much.

**Comment by shawnz** [comment]
It seems like PR 148 implements a similar solution but with less duplication

**Comment by zemian** [comment]
Hello @CShawn thank you so much for interested in Quartz project and providing this PR. Sorry it took us so long to respond! 

As you have noticed that this PR actual duplicate the work already submitted in #148. Given the other PR has less redundant code, we will close this PR and proceed to work on other PR instead.

Again, thank you for your contribution!

**Pull request closed without merging**